File: 0012/solution.cpp
#include <gtest/gtest.h>

#include <array>
#include <cstdlib>
#include <string>

using namespace std;

const array<char, 7> romans = {'I', 'V', 'X', 'L', 'C', 'D', 'M'};

class Solution {
 public:
  string intToRoman(int num) {
    string ret;
    size_t idx = 0;
    while (num > 0) {
      string digit;
      int remainder = num % 10;
      if (remainder == 4) {
        digit += romans[idx];
        digit += romans[idx+1];
      } else if (remainder == 9) {
        digit += romans[idx];
        digit += romans[idx+2];
      } else {
        if (remainder >= 5) {
          digit += romans[idx+1];
          remainder -= 5;
        }
        while (remainder > 0) {
          digit += romans[idx];
          remainder--;
        }
      }
      num /= 10;
      ret = digit + ret;
      idx += 2;
    }
    return ret;
  }
};

int main() {
  Solution sol;
  int num;
  string expected;

  num = 3, expected = "III";
  EXPECT_EQ(expected, sol.intToRoman(num));

  num = 4, expected = "IV";
  EXPECT_EQ(expected, sol.intToRoman(num));

  num = 9, expected = "IX";
  EXPECT_EQ(expected, sol.intToRoman(num));

  num = 58, expected = "LVIII";
  EXPECT_EQ(expected, sol.intToRoman(num));

  num = 1994, expected = "MCMXCIV";
  EXPECT_EQ(expected, sol.intToRoman(num));

  exit(EXIT_SUCCESS);
}

File: 0012/solution.py
from typing import *


class Solution:
    def intToRoman(self, num: int) -> str:
        mapping = [("", "", "M", 1000), ("M", "D", "C", 100), ("C", "L", "X", 10), ("X", "V", "I", 1)]
        chars = []
        for char10, char5, char1, value in mapping:
            quotient = int(num / value)
            num -= quotient * value
            if quotient == 9:
                chars.append(char1)
                chars.append(char10)
            elif quotient >= 5:
                chars.append(char5)
                for _ in range(quotient - 5):
                    chars.append(char1)
            elif quotient == 4:
                chars.append(char1)
                chars.append(char5)
            else:
                for _ in range(quotient):
                    chars.append(char1)
        return "".join(chars)


if __name__ == '__main__':
    sol = Solution()
    assert sol.intToRoman(3749) == "MMMDCCXLIX"
    assert sol.intToRoman(58) == "LVIII"
    assert sol.intToRoman(1994) == "MCMXCIV"
